src/ch341a.c

```diff
@@ -0,0 +1,467 @@
+    res[i] = pkt[i];
+  }
+  return size;
+}
+
+static int ch341a_spi_cmd(const PROGRAMMER *pgm, const unsigned char *cmd,
+                          unsigned char *res)
+{
+  return pgm->spi(pgm, cmd, res, 4);
+}
+
+static int ch341a_spi(const PROGRAMMER *pgm, const unsigned char *in,
+                      unsigned char *out, int size)
+{
+  unsigned char pkt[CH341A_PACKET_LENGTH];
+  unsigned int i;
+
+  if (!size) return 0;
+
+  if (size > CH341A_PACKET_LENGTH - 1) size = CH341A_PACKET_LENGTH - 1;
+
+  pkt[0] = CH341A_CMD_SPI_STREAM;
+
+  for (i = 0; i < size; i++) pkt[i + 1] = swap_byte(in[i]);
+
+  if (!CH341USBTransfer(pgm, LIBUSB_ENDPOINT_OUT, pkt, size + 1)) {
+    fprintf(stderr, "Error: failed to transfer data to CH341\n");
+    return -1;
+  }
+
+  if (!CH341USBTransfer(pgm, LIBUSB_ENDPOINT_IN, pkt, size)) {
+    fprintf(stderr, "Error: failed to transfer data from CH341\n");
+    return -1;
+  }
+
+  for (i = 0; i < size; i++) out[i] = swap_byte(pkt[i]);
+
+  return size;
+}
+
+static int ch341a_spi_program_enable(const PROGRAMMER *pgm, const AVRPART *p)
+{
+  unsigned char res[4];
+  unsigned char cmd[4];
+  memset(cmd, 0, sizeof(cmd));
+  memset(res, 0, sizeof(res));
+
+  cmd[0] = 0;
+  pmsg_trace("ch341a_program_enable() %s\n", p->op[AVR_OP_PGM_ENABLE]);
+
+  if (p->op[AVR_OP_PGM_ENABLE] == NULL) {
+    pmsg_error("program enable instruction not defined for part \"%s\"\n",
+              p->desc);
+    return -1;
+  }
+  avr_set_bits(p->op[AVR_OP_PGM_ENABLE], cmd);
+  pgm->cmd(pgm, cmd, res);
+
+  pmsg_debug("program_enable(): sending command. Resp = %x %x %x %x \n",
+            (int)res[0], (int)res[1], (int)res[2], (int)res[3]);
+  // check for sync character
+  if (res[2] != cmd[1]) return -2;
+  return 0;
+}
+
+static int ch341a_spi_chip_erase(const struct programmer_t *pgm,
+                                 const AVRPART *p)
+{
+  unsigned char cmd[4];
+  unsigned char res[4];
+  if (p->op[AVR_OP_CHIP_ERASE] == NULL) {
+    pmsg_error("chip erase instruction not defined for part \"%s\"\n", p->desc);
+    return -1;
+  }
+  memset(cmd, 0, sizeof(cmd));
+  avr_set_bits(p->op[AVR_OP_CHIP_ERASE], cmd);
+  pgm->cmd(pgm, cmd, res);
+  usleep(p->chip_erase_delay);
+  pgm->initialize(pgm, p);
+  return 0;
+}
+
+
+void ch341a_initpgm(PROGRAMMER * pgm)
+{
+  strcpy(pgm->type, "ch341a");
+  /*
+  * mandatory functions
+  */
+  pgm->initialize     = ch341a_initialize;
+  pgm->display        = ch341a_display;
+  pgm->enable         = ch341a_enable;
+  pgm->disable        = ch341a_disable;
+  pgm->program_enable = ch341a_spi_program_enable;
+  pgm->chip_erase     = ch341a_spi_chip_erase;
+  pgm->cmd            = ch341a_spi_cmd;
+  pgm->spi            = ch341a_spi;
+  pgm->open           = ch341a_open;
+  pgm->close          = ch341a_close;
+  pgm->read_byte      = avr_read_byte_default;
+  pgm->write_byte     = avr_write_byte_default;
+
+  /*
+  * optional functions
+  */
+
+  //pgm->paged_write    = ch341a_spi_paged_write;
```

Comment:
If the paged read/write routines are not supported then terminal cache does not work. 

@MCUdude and @mcuee Could you try, please, to uncomment these two lines and provide the following "fall-back" functions here
```
static int ch341a_spi_paged_write(const PROGRAMMER *pgm, const AVRPART *p, const AVRMEM *m,
  unsigned int page_size, unsigned int addr, unsigned int n_bytes) {
  
  if(n_bytes) {
    // Paged writes only valid for flash and eeprom
    if(!avr_mem_is_flash_type(m) && !avr_mem_is_eeprom_type(m))
      return -2;
  
    // Fall back on bytewise write
    for(unsigned int end = addr + n_bytes; addr < end; addr++)
      if(pgm->write_byte(pgm, p, m, addr, m->buf[addr]) < 0)
        return -1;
  }

  return n_bytes;
}
  

static int ch341a_spi_paged_load(const PROGRAMMER *pgm, const AVRPART *p, const AVRMEM *m,
  unsigned int page_size, unsigned int addr, unsigned int n_bytes) {
    
  if(n_bytes) { 
    // Paged reads only valid for flash and eeprom
    if(!avr_mem_is_flash_type(m) && !avr_mem_is_eeprom_type(m))
      return -2;

    // Fall back on bytewise read
    for(unsigned int end = addr + n_bytes; addr < end; addr++)
      if(pgm->read_byte(pgm, p, m, addr, m->buf+addr) < 0)
        return -1;
  }

  return n_bytes;
}
```

I haven't tried them (they might not even compile) but you get the gist.

Reply:
With these functions I'm able to write to EEPROM using cache write, but it fails to write to flash:

```
avrdude> write flash 0x10 "Hello World"
>>> write flash 0x10 "Hello World" 

Caching | ################################################## | 100% 0.00 s 

avrdude> flush
>>> flush 
avrdude: synching cache to device ... 
Writing | -------------------------------------------------- | 0% 0.04 s 

avrdude error: verification mismatch at flash page addr 0x0000
avrdude> quit
>>> quit 
```

Reply:
Any more intelligence on why this fails? More verbose output? Is the writing at fault or the reading back?  Is a corresponding `-U` call able to deal with this? ...

Reply:
Here is the `-vvvv` output:

```
$ ./avrdude -cch341a -patmega16 -t -vvvv
executable_abspath = /Users/hans/Downloads/avrdude-terminal/src/avrdude
executable_abspath_len = 50
executable_dirpath = /Users/hans/Downloads/avrdude-terminal/src
executable_dirpath_len = 42
sys_config = /Users/hans/Downloads/avrdude-terminal/src/avrdude.conf
sys_config_found = true


avrdude: Version 7.0-20220508
         Copyright the AVRDUDE authors;
         see https://github.com/avrdudes/avrdude/blob/main/AUTHORS

         System wide configuration file is /Users/hans/Downloads/avrdude-terminal/src/avrdude.conf
         User configuration file is /Users/hans/.avrduderc

         Using Port                    : usb
         Using Programmer              : ch341a
avrdude: ch341a_open("usb")
         AVR Part                      : ATmega16
         Chip Erase delay              : 9000 us
         PAGEL                         : PD7
         BS2                           : PA0
         RESET disposition             : possible i/o
         RETRY pulse                   : SCK
         Serial program mode           : yes
         Parallel program mode         : yes
         Timeout                       : 200
         StabDelay                     : 100
         CmdexeDelay                   : 25
         SyncLoops                     : 32
         PollIndex                     : 3
         PollValue                     : 0x53
         Memory Detail                 :

                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           eeprom                  4    10   128    0 no        512    4      0  9000  9000 0xff 0xff
                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           flash                  33     6   128    0 yes     16384  128    128  4500  4500 0xff 0xff
                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           lfuse                   0     0     0    0 no          1    1      0  9000  9000 0x00 0x00
                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           hfuse                   0     0     0    0 no          1    1      0  9000  9000 0x00 0x00
                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           lock                    0     0     0    0 no          1    1      0  9000  9000 0x00 0x00
                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           signature               0     0     0    0 no          3    1      0     0     0 0x00 0x00
                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           calibration             0     0     0    0 no          4    1      0     0     0 0x00 0x00

         Programmer Type : ch341a
         Description     : ch341a programmer

avrdude: ch341a_initialize()
avrdude: ch341a_ChipSelect()
avrdude: ch341a_ChipSelect()
avrdude: ch341a_program_enable() 
avrdude: program_enable(): sending command. Resp = ff ff 53 0 
avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.00 s 

avrdude: device signature = 0x1e9403 (probably m16)
avrdude> write flash 0 "Hello World"
>>> write flash 0 "Hello World" 
avrdude: (write) writing 12 bytes starting from address 0x00

Caching | ################################################## | 100% 0.02 s 

avrdude> flush
>>> flush 
avrdude: synching cache to device ... 
Reading | ################################################## | 100% 2.25 s 
Erasing |                                                    | 0% 0.00 s avrdude: ch341a_initialize()
avrdude: ch341a_ChipSelect()
avrdude: ch341a_ChipSelect()
avrdude: ch341a_program_enable() 
avrdude: program_enable(): sending command. Resp = ff ff 53 0 
Erasing | ################################################## | 100% 0.04 s 
Writing | -------------------------------------------------- | 0% 0.05 s 

avrdude avr_flush_cache() [avrcache.c:543] error: verification mismatch at flash page addr 0x0000
avrdude> quit
>>> quit 
avrdude> avrdude: synching cache to device ... 
Reading | ################################################## | 100% 0.00 s 
Erasing |                                                    | 0% 0.00 s avrdude: ch341a_initialize()
avrdude: ch341a_ChipSelect()
avrdude: ch341a_ChipSelect()
avrdude: ch341a_program_enable() 
avrdude: program_enable(): sending command. Resp = ff ff 53 0 
Erasing | ################################################## | 100% 0.04 s 
Writing | -------------------------------------------------- | 0% 0.04 s 

avrdude avr_flush_cache() [avrcache.c:543] error: verification mismatch at flash page addr 0x0000
avrdude: ch341a_close()
avrdude: ch341a_ChipSelect()

avrdude done.  Thank you.
```

Reply:
@stefanrueger with `pgm->paged_write`and `pgm->paged_load` uncommented I'm not able to write to flash using `-U` either:

<details>

<summary>Writing to flash using -U</summary>

```
$ ./avrdude -cch341a -patmega16 -e -Uflash:w:0x55_256b.hex -vvvv
executable_abspath = /Users/hans/Downloads/avrdude-terminal/src/avrdude
executable_abspath_len = 50
executable_dirpath = /Users/hans/Downloads/avrdude-terminal/src
executable_dirpath_len = 42
sys_config = /Users/hans/Downloads/avrdude-terminal/src/avrdude.conf
sys_config_found = true


avrdude: Version 7.0-20220508
         Copyright the AVRDUDE authors;
         see https://github.com/avrdudes/avrdude/blob/main/AUTHORS

         System wide configuration file is /Users/hans/Downloads/avrdude-terminal/src/avrdude.conf
         User configuration file is /Users/hans/.avrduderc

         Using Port                    : usb
         Using Programmer              : ch341a
avrdude: ch341a_open("usb")
avrdude: input file 0x55_256b.hex auto detected as Intel Hex
         AVR Part                      : ATmega16
         Chip Erase delay              : 9000 us
         PAGEL                         : PD7
         BS2                           : PA0
         RESET disposition             : possible i/o
         RETRY pulse                   : SCK
         Serial program mode           : yes
         Parallel program mode         : yes
         Timeout                       : 200
         StabDelay                     : 100
         CmdexeDelay                   : 25
         SyncLoops                     : 32
         PollIndex                     : 3
         PollValue                     : 0x53
         Memory Detail                 :

                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           eeprom                  4    10   128    0 no        512    4      0  9000  9000 0xff 0xff
                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           flash                  33     6   128    0 yes     16384  128    128  4500  4500 0xff 0xff
                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           lfuse                   0     0     0    0 no          1    1      0  9000  9000 0x00 0x00
                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           hfuse                   0     0     0    0 no          1    1      0  9000  9000 0x00 0x00
                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           lock                    0     0     0    0 no          1    1      0  9000  9000 0x00 0x00
                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           signature               0     0     0    0 no          3    1      0     0     0 0x00 0x00
                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           calibration             0     0     0    0 no          4    1      0     0     0 0x00 0x00

         Programmer Type : ch341a
         Description     : ch341a programmer

avrdude: ch341a_initialize()
avrdude: ch341a_ChipSelect()
avrdude: ch341a_ChipSelect()
avrdude: ch341a_program_enable() 
avrdude: program_enable(): sending command. Resp = ff ff 53 0 
avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.00 s 

avrdude: device signature = 0x1e9403 (probably m16)
avrdude: erasing chip
avrdude: ch341a_initialize()
avrdude: ch341a_ChipSelect()
avrdude: ch341a_ChipSelect()
avrdude: ch341a_program_enable() 
avrdude: program_enable(): sending command. Resp = ff ff 53 0 
avrdude: reading input file 0x55_256b.hex for flash
         with 256 bytes in 1 section within [0, 0xff]
         using 2 pages and 0 pad bytes
avrdude: writing 256 bytes flash ...

Writing | ################################################## | 100% 0.07 s 

avrdude: 256 bytes of flash written
avrdude: verifying flash memory against 0x55_256b.hex
avrdude: reading on-chip flash data ...

Reading | ################################################## | 100% 0.05 s 

avrdude: avr_read_mem(): skipping page 2: no interesting data
avrdude: avr_read_mem(): skipping page 3: no interesting data
avrdude: avr_read_mem(): skipping page 4: no interesting data
avrdude: avr_read_mem(): skipping page 5: no interesting data
avrdude: avr_read_mem(): skipping page 6: no interesting data
avrdude: avr_read_mem(): skipping page 7: no interesting data
avrdude: avr_read_mem(): skipping page 8: no interesting data
avrdude: avr_read_mem(): skipping page 9: no interesting data
avrdude: avr_read_mem(): skipping page 10: no interesting data
avrdude: avr_read_mem(): skipping page 11: no interesting data
avrdude: avr_read_mem(): skipping page 12: no interesting data
avrdude: avr_read_mem(): skipping page 13: no interesting data
avrdude: avr_read_mem(): skipping page 14: no interesting data
avrdude: avr_read_mem(): skipping page 15: no interesting data
avrdude: avr_read_mem(): skipping page 16: no interesting data
avrdude: avr_read_mem(): skipping page 17: no interesting data
avrdude: avr_read_mem(): skipping page 18: no interesting data
avrdude: avr_read_mem(): skipping page 19: no interesting data
avrdude: avr_read_mem(): skipping page 20: no interesting data
avrdude: avr_read_mem(): skipping page 21: no interesting data
avrdude: avr_read_mem(): skipping page 22: no interesting data
avrdude: avr_read_mem(): skipping page 23: no interesting data
avrdude: avr_read_mem(): skipping page 24: no interesting data
avrdude: avr_read_mem(): skipping page 25: no interesting data
avrdude: avr_read_mem(): skipping page 26: no interesting data
avrdude: avr_read_mem(): skipping page 27: no interesting data
avrdude: avr_read_mem(): skipping page 28: no interesting data
avrdude: avr_read_mem(): skipping page 29: no interesting data
avrdude: avr_read_mem(): skipping page 30: no interesting data
avrdude: avr_read_mem(): skipping page 31: no interesting data
avrdude: avr_read_mem(): skipping page 32: no interesting data
avrdude: avr_read_mem(): skipping page 33: no interesting data
avrdude: avr_read_mem(): skipping page 34: no interesting data
avrdude: avr_read_mem(): skipping page 35: no interesting data
avrdude: avr_read_mem(): skipping page 36: no interesting data
avrdude: avr_read_mem(): skipping page 37: no interesting data
avrdude: avr_read_mem(): skipping page 38: no interesting data
avrdude: avr_read_mem(): skipping page 39: no interesting data
avrdude: avr_read_mem(): skipping page 40: no interesting data
avrdude: avr_read_mem(): skipping page 41: no interesting data
avrdude: avr_read_mem(): skipping page 42: no interesting data
avrdude: avr_read_mem(): skipping page 43: no interesting data
avrdude: avr_read_mem(): skipping page 44: no interesting data
avrdude: avr_read_mem(): skipping page 45: no interesting data
avrdude: avr_read_mem(): skipping page 46: no interesting data
avrdude: avr_read_mem(): skipping page 47: no interesting data
avrdude: avr_read_mem(): skipping page 48: no interesting data
avrdude: avr_read_mem(): skipping page 49: no interesting data
avrdude: avr_read_mem(): skipping page 50: no interesting data
avrdude: avr_read_mem(): skipping page 51: no interesting data
avrdude: avr_read_mem(): skipping page 52: no interesting data
avrdude: avr_read_mem(): skipping page 53: no interesting data
avrdude: avr_read_mem(): skipping page 54: no interesting data
avrdude: avr_read_mem(): skipping page 55: no interesting data
avrdude: avr_read_mem(): skipping page 56: no interesting data
avrdude: avr_read_mem(): skipping page 57: no interesting data
avrdude: avr_read_mem(): skipping page 58: no interesting data
avrdude: avr_read_mem(): skipping page 59: no interesting data
avrdude: avr_read_mem(): skipping page 60: no interesting data
avrdude: avr_read_mem(): skipping page 61: no interesting data
avrdude: avr_read_mem(): skipping page 62: no interesting data
avrdude: avr_read_mem(): skipping page 63: no interesting data
avrdude: avr_read_mem(): skipping page 64: no interesting data
avrdude: avr_read_mem(): skipping page 65: no interesting data
avrdude: avr_read_mem(): skipping page 66: no interesting data
avrdude: avr_read_mem(): skipping page 67: no interesting data
avrdude: avr_read_mem(): skipping page 68: no interesting data
avrdude: avr_read_mem(): skipping page 69: no interesting data
avrdude: avr_read_mem(): skipping page 70: no interesting data
avrdude: avr_read_mem(): skipping page 71: no interesting data
avrdude: avr_read_mem(): skipping page 72: no interesting data
avrdude: avr_read_mem(): skipping page 73: no interesting data
avrdude: avr_read_mem(): skipping page 74: no interesting data
avrdude: avr_read_mem(): skipping page 75: no interesting data
avrdude: avr_read_mem(): skipping page 76: no interesting data
avrdude: avr_read_mem(): skipping page 77: no interesting data
avrdude: avr_read_mem(): skipping page 78: no interesting data
avrdude: avr_read_mem(): skipping page 79: no interesting data
avrdude: avr_read_mem(): skipping page 80: no interesting data
avrdude: avr_read_mem(): skipping page 81: no interesting data
avrdude: avr_read_mem(): skipping page 82: no interesting data
avrdude: avr_read_mem(): skipping page 83: no interesting data
avrdude: avr_read_mem(): skipping page 84: no interesting data
avrdude: avr_read_mem(): skipping page 85: no interesting data
avrdude: avr_read_mem(): skipping page 86: no interesting data
avrdude: avr_read_mem(): skipping page 87: no interesting data
avrdude: avr_read_mem(): skipping page 88: no interesting data
avrdude: avr_read_mem(): skipping page 89: no interesting data
avrdude: avr_read_mem(): skipping page 90: no interesting data
avrdude: avr_read_mem(): skipping page 91: no interesting data
avrdude: avr_read_mem(): skipping page 92: no interesting data
avrdude: avr_read_mem(): skipping page 93: no interesting data
avrdude: avr_read_mem(): skipping page 94: no interesting data
avrdude: avr_read_mem(): skipping page 95: no interesting data
avrdude: avr_read_mem(): skipping page 96: no interesting data
avrdude: avr_read_mem(): skipping page 97: no interesting data
avrdude: avr_read_mem(): skipping page 98: no interesting data
avrdude: avr_read_mem(): skipping page 99: no interesting data
avrdude: avr_read_mem(): skipping page 100: no interesting data
avrdude: avr_read_mem(): skipping page 101: no interesting data
avrdude: avr_read_mem(): skipping page 102: no interesting data
avrdude: avr_read_mem(): skipping page 103: no interesting data
avrdude: avr_read_mem(): skipping page 104: no interesting data
avrdude: avr_read_mem(): skipping page 105: no interesting data
avrdude: avr_read_mem(): skipping page 106: no interesting data
avrdude: avr_read_mem(): skipping page 107: no interesting data
avrdude: avr_read_mem(): skipping page 108: no interesting data
avrdude: avr_read_mem(): skipping page 109: no interesting data
avrdude: avr_read_mem(): skipping page 110: no interesting data
avrdude: avr_read_mem(): skipping page 111: no interesting data
avrdude: avr_read_mem(): skipping page 112: no interesting data
avrdude: avr_read_mem(): skipping page 113: no interesting data
avrdude: avr_read_mem(): skipping page 114: no interesting data
avrdude: avr_read_mem(): skipping page 115: no interesting data
avrdude: avr_read_mem(): skipping page 116: no interesting data
avrdude: avr_read_mem(): skipping page 117: no interesting data
avrdude: avr_read_mem(): skipping page 118: no interesting data
avrdude: avr_read_mem(): skipping page 119: no interesting data
avrdude: avr_read_mem(): skipping page 120: no interesting data
avrdude: avr_read_mem(): skipping page 121: no interesting data
avrdude: avr_read_mem(): skipping page 122: no interesting data
avrdude: avr_read_mem(): skipping page 123: no interesting data
avrdude: avr_read_mem(): skipping page 124: no interesting data
avrdude: avr_read_mem(): skipping page 125: no interesting data
avrdude: avr_read_mem(): skipping page 126: no interesting data
avrdude: avr_read_mem(): skipping page 127: no interesting data
avrdude: verifying ...
avrdude avr_verify() [avr.c:1234] warning: verification mismatch
        device 0xff != input 0x55 at addr 0x0000 (error)
        device 0xff != input 0x55 at addr 0x0001 (error)
        device 0xff != input 0x55 at addr 0x0002 (error)
        device 0xff != input 0x55 at addr 0x0003 (error)
        device 0xff != input 0x55 at addr 0x0004 (error)
        device 0xff != input 0x55 at addr 0x0005 (error)
        device 0xff != input 0x55 at addr 0x0006 (error)
        device 0xff != input 0x55 at addr 0x0007 (error)
        device 0xff != input 0x55 at addr 0x0008 (error)
        device 0xff != input 0x55 at addr 0x0009 (error)
        device 0xff != input 0x55 at addr 0x000a (error)
        device 0xff != input 0x55 at addr 0x000b (error)
        device 0xff != input 0x55 at addr 0x000c (error)
        device 0xff != input 0x55 at addr 0x000d (error)
        device 0xff != input 0x55 at addr 0x000e (error)
        device 0xff != input 0x55 at addr 0x000f (error)
        device 0xff != input 0x55 at addr 0x0010 (error)
        device 0xff != input 0x55 at addr 0x0011 (error)
        device 0xff != input 0x55 at addr 0x0012 (error)
        device 0xff != input 0x55 at addr 0x0013 (error)
        device 0xff != input 0x55 at addr 0x0014 (error)
        device 0xff != input 0x55 at addr 0x0015 (error)
        device 0xff != input 0x55 at addr 0x0016 (error)
        device 0xff != input 0x55 at addr 0x0017 (error)
        device 0xff != input 0x55 at addr 0x0018 (error)
        device 0xff != input 0x55 at addr 0x0019 (error)
        device 0xff != input 0x55 at addr 0x001a (error)
        device 0xff != input 0x55 at addr 0x001b (error)
        device 0xff != input 0x55 at addr 0x001c (error)
        device 0xff != input 0x55 at addr 0x001d (error)
        device 0xff != input 0x55 at addr 0x001e (error)
        device 0xff != input 0x55 at addr 0x001f (error)
        device 0xff != input 0x55 at addr 0x0020 (error)
        device 0xff != input 0x55 at addr 0x0021 (error)
        device 0xff != input 0x55 at addr 0x0022 (error)
        device 0xff != input 0x55 at addr 0x0023 (error)
        device 0xff != input 0x55 at addr 0x0024 (error)
        device 0xff != input 0x55 at addr 0x0025 (error)
        device 0xff != input 0x55 at addr 0x0026 (error)
        device 0xff != input 0x55 at addr 0x0027 (error)
        device 0xff != input 0x55 at addr 0x0028 (error)
        device 0xff != input 0x55 at addr 0x0029 (error)
        device 0xff != input 0x55 at addr 0x002a (error)
        device 0xff != input 0x55 at addr 0x002b (error)
        device 0xff != input 0x55 at addr 0x002c (error)
        device 0xff != input 0x55 at addr 0x002d (error)
        device 0xff != input 0x55 at addr 0x002e (error)
        device 0xff != input 0x55 at addr 0x002f (error)
        device 0xff != input 0x55 at addr 0x0030 (error)
        device 0xff != input 0x55 at addr 0x0031 (error)
        device 0xff != input 0x55 at addr 0x0032 (error)
        device 0xff != input 0x55 at addr 0x0033 (error)
        device 0xff != input 0x55 at addr 0x0034 (error)
        device 0xff != input 0x55 at addr 0x0035 (error)
        device 0xff != input 0x55 at addr 0x0036 (error)
        device 0xff != input 0x55 at addr 0x0037 (error)
        device 0xff != input 0x55 at addr 0x0038 (error)
        device 0xff != input 0x55 at addr 0x0039 (error)
        device 0xff != input 0x55 at addr 0x003a (error)
        device 0xff != input 0x55 at addr 0x003b (error)
        device 0xff != input 0x55 at addr 0x003c (error)
        device 0xff != input 0x55 at addr 0x003d (error)
        device 0xff != input 0x55 at addr 0x003e (error)
        device 0xff != input 0x55 at addr 0x003f (error)
        device 0xff != input 0x55 at addr 0x0040 (error)
        device 0xff != input 0x55 at addr 0x0041 (error)
        device 0xff != input 0x55 at addr 0x0042 (error)
        device 0xff != input 0x55 at addr 0x0043 (error)
        device 0xff != input 0x55 at addr 0x0044 (error)
        device 0xff != input 0x55 at addr 0x0045 (error)
        device 0xff != input 0x55 at addr 0x0046 (error)
        device 0xff != input 0x55 at addr 0x0047 (error)
        device 0xff != input 0x55 at addr 0x0048 (error)
        device 0xff != input 0x55 at addr 0x0049 (error)
        device 0xff != input 0x55 at addr 0x004a (error)
        device 0xff != input 0x55 at addr 0x004b (error)
        device 0xff != input 0x55 at addr 0x004c (error)
        device 0xff != input 0x55 at addr 0x004d (error)
        device 0xff != input 0x55 at addr 0x004e (error)
        device 0xff != input 0x55 at addr 0x004f (error)
        device 0xff != input 0x55 at addr 0x0050 (error)
        device 0xff != input 0x55 at addr 0x0051 (error)
        device 0xff != input 0x55 at addr 0x0052 (error)
        device 0xff != input 0x55 at addr 0x0053 (error)
        device 0xff != input 0x55 at addr 0x0054 (error)
        device 0xff != input 0x55 at addr 0x0055 (error)
        device 0xff != input 0x55 at addr 0x0056 (error)
        device 0xff != input 0x55 at addr 0x0057 (error)
        device 0xff != input 0x55 at addr 0x0058 (error)
        device 0xff != input 0x55 at addr 0x0059 (error)
        device 0xff != input 0x55 at addr 0x005a (error)
        device 0xff != input 0x55 at addr 0x005b (error)
        device 0xff != input 0x55 at addr 0x005c (error)
        device 0xff != input 0x55 at addr 0x005d (error)
        device 0xff != input 0x55 at addr 0x005e (error)
        device 0xff != input 0x55 at addr 0x005f (error)
        device 0xff != input 0x55 at addr 0x0060 (error)
        device 0xff != input 0x55 at addr 0x0061 (error)
        device 0xff != input 0x55 at addr 0x0062 (error)
        device 0xff != input 0x55 at addr 0x0063 (error)
        device 0xff != input 0x55 at addr 0x0064 (error)
        device 0xff != input 0x55 at addr 0x0065 (error)
        device 0xff != input 0x55 at addr 0x0066 (error)
        device 0xff != input 0x55 at addr 0x0067 (error)
        device 0xff != input 0x55 at addr 0x0068 (error)
        device 0xff != input 0x55 at addr 0x0069 (error)
        device 0xff != input 0x55 at addr 0x006a (error)
        device 0xff != input 0x55 at addr 0x006b (error)
        device 0xff != input 0x55 at addr 0x006c (error)
        device 0xff != input 0x55 at addr 0x006d (error)
        device 0xff != input 0x55 at addr 0x006e (error)
        device 0xff != input 0x55 at addr 0x006f (error)
        device 0xff != input 0x55 at addr 0x0070 (error)
        device 0xff != input 0x55 at addr 0x0071 (error)
        device 0xff != input 0x55 at addr 0x0072 (error)
        device 0xff != input 0x55 at addr 0x0073 (error)
        device 0xff != input 0x55 at addr 0x0074 (error)
        device 0xff != input 0x55 at addr 0x0075 (error)
        device 0xff != input 0x55 at addr 0x0076 (error)
        device 0xff != input 0x55 at addr 0x0077 (error)
        device 0xff != input 0x55 at addr 0x0078 (error)
        device 0xff != input 0x55 at addr 0x0079 (error)
        device 0xff != input 0x55 at addr 0x007a (error)
        device 0xff != input 0x55 at addr 0x007b (error)
        device 0xff != input 0x55 at addr 0x007c (error)
        device 0xff != input 0x55 at addr 0x007d (error)
        device 0xff != input 0x55 at addr 0x007e (error)
        device 0xff != input 0x55 at addr 0x007f (error)
        device 0xff != input 0x55 at addr 0x0080 (error)
        device 0xff != input 0x55 at addr 0x0081 (error)
        device 0xff != input 0x55 at addr 0x0082 (error)
        device 0xff != input 0x55 at addr 0x0083 (error)
        device 0xff != input 0x55 at addr 0x0084 (error)
        device 0xff != input 0x55 at addr 0x0085 (error)
        device 0xff != input 0x55 at addr 0x0086 (error)
        device 0xff != input 0x55 at addr 0x0087 (error)
        device 0xff != input 0x55 at addr 0x0088 (error)
        device 0xff != input 0x55 at addr 0x0089 (error)
        device 0xff != input 0x55 at addr 0x008a (error)
        device 0xff != input 0x55 at addr 0x008b (error)
        device 0xff != input 0x55 at addr 0x008c (error)
        device 0xff != input 0x55 at addr 0x008d (error)
        device 0xff != input 0x55 at addr 0x008e (error)
        device 0xff != input 0x55 at addr 0x008f (error)
        device 0xff != input 0x55 at addr 0x0090 (error)
        device 0xff != input 0x55 at addr 0x0091 (error)
        device 0xff != input 0x55 at addr 0x0092 (error)
        device 0xff != input 0x55 at addr 0x0093 (error)
        device 0xff != input 0x55 at addr 0x0094 (error)
        device 0xff != input 0x55 at addr 0x0095 (error)
        device 0xff != input 0x55 at addr 0x0096 (error)
        device 0xff != input 0x55 at addr 0x0097 (error)
        device 0xff != input 0x55 at addr 0x0098 (error)
        device 0xff != input 0x55 at addr 0x0099 (error)
        device 0xff != input 0x55 at addr 0x009a (error)
        device 0xff != input 0x55 at addr 0x009b (error)
        device 0xff != input 0x55 at addr 0x009c (error)
        device 0xff != input 0x55 at addr 0x009d (error)
        device 0xff != input 0x55 at addr 0x009e (error)
        device 0xff != input 0x55 at addr 0x009f (error)
        device 0xff != input 0x55 at addr 0x00a0 (error)
        device 0xff != input 0x55 at addr 0x00a1 (error)
        device 0xff != input 0x55 at addr 0x00a2 (error)
        device 0xff != input 0x55 at addr 0x00a3 (error)
        device 0xff != input 0x55 at addr 0x00a4 (error)
        device 0xff != input 0x55 at addr 0x00a5 (error)
        device 0xff != input 0x55 at addr 0x00a6 (error)
        device 0xff != input 0x55 at addr 0x00a7 (error)
        device 0xff != input 0x55 at addr 0x00a8 (error)
        device 0xff != input 0x55 at addr 0x00a9 (error)
        device 0xff != input 0x55 at addr 0x00aa (error)
        device 0xff != input 0x55 at addr 0x00ab (error)
        device 0xff != input 0x55 at addr 0x00ac (error)
        device 0xff != input 0x55 at addr 0x00ad (error)
        device 0xff != input 0x55 at addr 0x00ae (error)
        device 0xff != input 0x55 at addr 0x00af (error)
        device 0xff != input 0x55 at addr 0x00b0 (error)
        device 0xff != input 0x55 at addr 0x00b1 (error)
        device 0xff != input 0x55 at addr 0x00b2 (error)
        device 0xff != input 0x55 at addr 0x00b3 (error)
        device 0xff != input 0x55 at addr 0x00b4 (error)
        device 0xff != input 0x55 at addr 0x00b5 (error)
        device 0xff != input 0x55 at addr 0x00b6 (error)
        device 0xff != input 0x55 at addr 0x00b7 (error)
        device 0xff != input 0x55 at addr 0x00b8 (error)
        device 0xff != input 0x55 at addr 0x00b9 (error)
        device 0xff != input 0x55 at addr 0x00ba (error)
        device 0xff != input 0x55 at addr 0x00bb (error)
        device 0xff != input 0x55 at addr 0x00bc (error)
        device 0xff != input 0x55 at addr 0x00bd (error)
        device 0xff != input 0x55 at addr 0x00be (error)
        device 0xff != input 0x55 at addr 0x00bf (error)
        device 0xff != input 0x55 at addr 0x00c0 (error)
        device 0xff != input 0x55 at addr 0x00c1 (error)
        device 0xff != input 0x55 at addr 0x00c2 (error)
        device 0xff != input 0x55 at addr 0x00c3 (error)
        device 0xff != input 0x55 at addr 0x00c4 (error)
        device 0xff != input 0x55 at addr 0x00c5 (error)
        device 0xff != input 0x55 at addr 0x00c6 (error)
        device 0xff != input 0x55 at addr 0x00c7 (error)
        device 0xff != input 0x55 at addr 0x00c8 (error)
        device 0xff != input 0x55 at addr 0x00c9 (error)
        device 0xff != input 0x55 at addr 0x00ca (error)
        device 0xff != input 0x55 at addr 0x00cb (error)
        device 0xff != input 0x55 at addr 0x00cc (error)
        device 0xff != input 0x55 at addr 0x00cd (error)
        device 0xff != input 0x55 at addr 0x00ce (error)
        device 0xff != input 0x55 at addr 0x00cf (error)
        device 0xff != input 0x55 at addr 0x00d0 (error)
        device 0xff != input 0x55 at addr 0x00d1 (error)
        device 0xff != input 0x55 at addr 0x00d2 (error)
        device 0xff != input 0x55 at addr 0x00d3 (error)
        device 0xff != input 0x55 at addr 0x00d4 (error)
        device 0xff != input 0x55 at addr 0x00d5 (error)
        device 0xff != input 0x55 at addr 0x00d6 (error)
        device 0xff != input 0x55 at addr 0x00d7 (error)
        device 0xff != input 0x55 at addr 0x00d8 (error)
        device 0xff != input 0x55 at addr 0x00d9 (error)
        device 0xff != input 0x55 at addr 0x00da (error)
        device 0xff != input 0x55 at addr 0x00db (error)
        device 0xff != input 0x55 at addr 0x00dc (error)
        device 0xff != input 0x55 at addr 0x00dd (error)
        device 0xff != input 0x55 at addr 0x00de (error)
        device 0xff != input 0x55 at addr 0x00df (error)
        device 0xff != input 0x55 at addr 0x00e0 (error)
        device 0xff != input 0x55 at addr 0x00e1 (error)
        device 0xff != input 0x55 at addr 0x00e2 (error)
        device 0xff != input 0x55 at addr 0x00e3 (error)
        device 0xff != input 0x55 at addr 0x00e4 (error)
        device 0xff != input 0x55 at addr 0x00e5 (error)
        device 0xff != input 0x55 at addr 0x00e6 (error)
        device 0xff != input 0x55 at addr 0x00e7 (error)
        device 0xff != input 0x55 at addr 0x00e8 (error)
        device 0xff != input 0x55 at addr 0x00e9 (error)
        device 0xff != input 0x55 at addr 0x00ea (error)
        device 0xff != input 0x55 at addr 0x00eb (error)
        device 0xff != input 0x55 at addr 0x00ec (error)
        device 0xff != input 0x55 at addr 0x00ed (error)
        device 0xff != input 0x55 at addr 0x00ee (error)
        device 0xff != input 0x55 at addr 0x00ef (error)
        device 0xff != input 0x55 at addr 0x00f0 (error)
        device 0xff != input 0x55 at addr 0x00f1 (error)
        device 0xff != input 0x55 at addr 0x00f2 (error)
        device 0xff != input 0x55 at addr 0x00f3 (error)
        device 0xff != input 0x55 at addr 0x00f4 (error)
        device 0xff != input 0x55 at addr 0x00f5 (error)
        device 0xff != input 0x55 at addr 0x00f6 (error)
        device 0xff != input 0x55 at addr 0x00f7 (error)
        device 0xff != input 0x55 at addr 0x00f8 (error)
        device 0xff != input 0x55 at addr 0x00f9 (error)
        device 0xff != input 0x55 at addr 0x00fa (error)
        device 0xff != input 0x55 at addr 0x00fb (error)
        device 0xff != input 0x55 at addr 0x00fc (error)
        device 0xff != input 0x55 at addr 0x00fd (error)
        device 0xff != input 0x55 at addr 0x00fe (error)
        device 0xff != input 0x55 at addr 0x00ff (error)
avrdude do_op() [update.c:620] error: verification mismatch
avrdude: ch341a_close()
avrdude: ch341a_ChipSelect()

avrdude done.  Thank you.
```

</details>

Reply:
But does it write OK w/out paged function?!? That would be ... interesting

Reply:
... and you have recompiled all from scratch (rather than relying on dates and make)? Just checking, ...

Reply:
> But does it write OK w/out paged function?!? That would be ... interesting
... and you have recompiled all from scratch (rather than relying on dates and make)? Just checking, ...

Yes, it is! and I've ran `make clean` to make sure the build is correct.

Reply:
Blimey. What an unexpected behaviour. Hmm, stabbing in the dark here, but wondering whether the paged version fails to wait after writing a page... perhaps insert  `usleep(20*1000);` before `return n_bytes;` near the end of `ch341a_spi_paged_write()`. Or sleep a few ms at the beginning of every `ch321a_spi_cmd()`? Just one hypothesis...

Reply:
Sorry, that didn't help either!

I tried the following approach. It doesn't work, but now it's also much slower:
```c
static int ch341a_spi_cmd(const PROGRAMMER *pgm, const unsigned char *cmd,
                          unsigned char *res)
{
  usleep(2000);
  return pgm->spi(pgm, cmd, res, 4);
}
```

Reply:
OK, nothing beats looking at the source code. I think we need to put an explicit avr_write_page() command into the paged write routine. I have refined the paged access functions as follows:

```
// Fall back on bytewise write (followed by write page if flash)
static int ch341a_spi_paged_write(const PROGRAMMER *pgm, const AVRPART *p, const AVRMEM *m,
  unsigned int page_size, unsigned int addr, unsigned int n_bytes) {
  
  int isflash = avr_mem_is_flash_type(m);
  
  if(n_bytes) {
    if(!isflash && !avr_mem_is_eeprom_type(m))  
      return -2;
      
    // Always called with addr at page boundary and n_bytes == m->page_size
    for(unsigned int end = addr + n_bytes; addr < end; addr++)
      if(pgm->write_byte(pgm, p, m, addr, m->buf[addr]) < 0) 
        return -1;
  }
    
  if(isflash && avr_write_page(pgm, p, m, addr-n_bytes) < 0)
    return -1;
  
  return n_bytes;
}  
 

// Fall back on bytewise read
static int ch341a_spi_paged_load(const PROGRAMMER *pgm, const AVRPART *p, const AVRMEM *m,
  unsigned int page_size, unsigned int addr, unsigned int n_bytes) {

  int isflash = avr_mem_is_flash_type(m);
  
  if(n_bytes) {
    if(!isflash && !avr_mem_is_eeprom_type(m))
      return -2;
    
    // Always called with addr at page boundary and n_bytes == m->page_size
    if(isflash && m->op[AVR_OP_LOAD_EXT_ADDR]) {
      unsigned char cmd[4], res[4];   
  
      memset(cmd, 0, sizeof cmd);
      avr_set_bits(m->op[AVR_OP_LOAD_EXT_ADDR], cmd);
      avr_set_addr(m->op[AVR_OP_LOAD_EXT_ADDR], cmd, addr/2);
      if(pgm->cmd(pgm, cmd, res) < 0)
        return -1;
   }
  
   for(unsigned int end = addr + n_bytes; addr < end; addr++)
     if(pgm->read_byte(pgm, p, m, addr, m->buf+addr) < 0)
       return -1;
  }
    
  return n_bytes;
}
```

Reply:
... if that works then it should also work for parts with large flash that need the load extended address cmd, eg, 2560

Reply:
Wow, this works perfectly! Excellent job @stefanrueger!

```
$ ./avrdude -cch341a -patmega16 -t

avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e9403 (probably m16)
avrdude> read flash
>>> read flash 

Reading | ################################################## | 100% 0.04 s 

0000  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0010  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0020  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0030  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0040  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0050  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0060  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0070  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0080  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0090  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
00a0  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
00b0  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
00c0  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
00d0  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
00e0  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
00f0  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|

avrdude> write flash 0 "Hello World!"
>>> write flash 0 "Hello World!" 

Caching | ################################################## | 100% 0.00 s 

avrdude> flush
>>> flush 
avrdude: synching cache to device ... 
Reading | ################################################## | 100% 2.24 s 
Erasing | ################################################## | 100% 0.04 s 
Writing | ################################################## | 100% 0.10 s 
avrdude> read flash 0 0x100
>>> read flash 0 0x100 

Reading | ################################################## | 100% 0.00 s 

0000  48 65 6c 6c 6f 20 57 6f  72 6c 64 21 00 55 55 55  |Hello World!.UUU|
0010  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0020  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0030  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0040  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0050  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0060  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0070  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0080  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
0090  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
00a0  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
00b0  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
00c0  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
00d0  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
00e0  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|
00f0  55 55 55 55 55 55 55 55  55 55 55 55 55 55 55 55  |UUUUUUUUUUUUUUUU|

avrdude> quit
>>> quit 
avrdude> 
avrdude done.  Thank you.
```

EEPROM r/w in terminal mode also works fine

Reply:
Thanks @MCUdude , that's great. So we have all in hand to make this PR work for AVRDUDE. Though I am sure @mcuee will try to break this programmer with fiendishly difficult flash files with holes below and above the 128k boundary for the ATmega2560 :)

How should we proceed? I can push a commit onto this PR with the paged access routines, and I can also correct the minor items that my review uncovered using the info that @mcuee provided. Would that be OK, @Windol @am-ar? I'd squash merge and credit everyone on this collaboration incl @Windol @am-ar @MCUdude and @mcuee.

Other suggestions anyone?